add castling logic to king model

erinmadisonanthony opened the pull request:
Migrated times_moved into the pieces model to track how many times each piece has been moved (and whether it is a piece's first move). Then added logic to the king model to recognize and allow castling. Still need to add the code to tell the front end to move the appropriate rook to the appropriate spot on the board.

hoitomt commented on app/models/king.rb:
```diff
@@ -7,6 +8,9 @@ def valid_moves
         coordinate
+      #check for castling  
+      elsif times_moved == 0 && (x_position - x).abs == 2 && (y_position - y) == 0
+        coordinate
```
Set a variable to indicate that casting is happening

```
@castling = true
coordinate
```

hoitomt commented on app/models/king.rb:
```diff
@@ -22,4 +26,9 @@ def move_path(destination)
   end
+
+  def castling?
+    return true if times_moved == 1 && (x_position == 2 || x_position == 6)
```
Update this to return `@castling`

erinmadisonanthony commented:
I added a valid_castling? method in the Game model to validate all of the conditions required for castling. I also updated the castling? method in the King model and added a castling? method in the Rook model. Please note that getting the King to stay in the castling position now requires a browser refresh, which I believe we already have a ticket for on Trello.

wookee55 commented:
Get this error running the code:
NoMethodError (undefined method `x_position' for #<Array:0xb5226540>):
  app/models/game.rb:122:in `valid_castling?'
  app/models/game.rb:72:in `move_piece'
  app/controllers/games_controller.rb:29:in `update'

Seems like invoked_rook is an array instead of a model object.

hoitomt commented on app/models/game.rb:
```diff
@@ -105,4 +108,53 @@ def is_obstructed?(destination)
+      #select which rook is castling
+      find_rook = pieces.select { |piece| piece.type == "Rook" && piece.user == @current_piece.user && (piece.x_position - @current_piece.x_position).abs <= 2 }
+      if !find_rook.empty?
+        invoked_rook = find_rook[0]
```
The variable `invoked_rook` is not available outside of this `if....end` conditional. You will need to make sure it is defined outside of this `if` statement.

Something like this would work:

```
invoked_rook = nil
find_rook = pieces.select { |piece| piece.type == "Rook" && piece.user == @current_piece.user && (piece.x_position - @current_piece.x_position).abs <= 2 }
if !find_rook.empty?
  invoked_rook = find_rook[0]
else
  return false
end
```

But it might be nicer to create a private method in this game model called `castling_rook`that looks like this:

```
def castling_rook
  pieces.select { |piece| piece.type == "Rook" && piece.user == @current_piece.user && (piece.x_position - @current_piece.x_position).abs <= 2 }.first
end
```

Then update the logic to be:
```
invoked_rook = castling_rook
return unless invoked_rook
```

hoitomt commented on app/models/game.rb:
```diff
@@ -105,4 +108,53 @@ def is_obstructed?(destination)
+    if @current_piece.type == "King" && @current_piece.castling?
+
+      #select which rook is castling
+      find_rook = pieces.select { |piece| piece.type == "Rook" && piece.user == @current_piece.user && (piece.x_position - @current_piece.x_position).abs <= 2 }
```
Is `@current_piece.x_position` the same thing as `destination`? If I'm understanding the logic correctly we want to evaluate the destination to determine if there is a Rook within 2 spaces of it. Is that correct?

hoitomt commented on app/models/game.rb:
```diff
@@ -105,4 +108,53 @@ def is_obstructed?(destination)
+          next if piece.captured?
+          return false if piece.x_position == 1 && piece.y_position == invoked_rook.y_position
+        end
+      end 
```
I don't think I understand what this code is doing. So here is our layout

Queen Side:

| 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 |
| --- | --- | --- | --- | --- | --- | --- | --- |
| | | | K | | | | R |

- I think the difference in the positions is 4 so it would be `@current_piece.x_position - invoked_rook.x_position == 4`
- We only need to look for pieces in positions 4, 5, and 6. At one time we had a method that we could give coordinates to and it would return the piece that was in that position. We got rid of that once we started using piece_id. But you can add it back in so that you can loop through positions (4, invoked_rook.y_position), (5, invoked_rook.y_position), and (7, invoked_rook.y_position) and call something like `Piece.where(x_position: x, y_position: y)`
- Return false if you find a piece in any of those positions.

Please let me know if I'm interpreting this incorrectly

hoitomt commented on app/models/game.rb:
```diff
@@ -105,4 +108,53 @@ def is_obstructed?(destination)
+            return false if attack_path.include?(space)
+          end
+        end
+      end
```
If I'm understanding this correctly - we are trying to make sure we don't castle into check, is that correct?  If so, we don't need to evaluate the king's move path, only it's final destination.

Also, we should definitely pull this logic out into it's own method. Optimally the method would take coordinates and return true or false if the coordinates lie on an attacking_pieces move path

```
def is_check?(coordinates)
  ...
end
```

hoitomt commented on app/models/rook.rb:
```diff
@@ -13,3 +13,7 @@ def valid_moves
   end
 end
+
+def castling?
```
I'm not sure if we need this. We may just want to depend on the King to tell us if we are castling. In your logic from the `game` class I don't think you ever call the Rook and ask it if is castling.

hoitomt commented on config/environments/development.rb:
```diff
@@ -39,4 +39,7 @@
   # config.action_view.raise_on_missing_translations = true
+
+  #fix config change preventing rails from running in local environment
+  config.web_console.whitelisted_ips = '10.0.2.2'
```
I haven't seen this before, what does this do?

hoitomt commented:
@erinmadisonanthony - I added some comments. Please let me know if it would be easier to get together in person to go over them.
